File: app/src/main/java/io/github/zwieback/familyfinance/core/model/ISmsPattern.java
package io.github.zwieback.familyfinance.core.model;

import android.databinding.Bindable;

import io.github.zwieback.familyfinance.core.model.restriction.SmsPatternRestriction;
import io.requery.CascadeAction;
import io.requery.Column;
import io.requery.Entity;
import io.requery.ManyToOne;
import io.requery.Nullable;
import io.requery.PropertyNameStyle;
import io.requery.Table;

@Entity(propertyNameStyle = PropertyNameStyle.FLUENT_BEAN)
@Table(name = "t_sms_pattern")
public interface ISmsPattern extends IBaseEntity {

    @Bindable
    @ManyToOne(cascade = CascadeAction.NONE)
    @Column(name = "template_id")
    ITemplate getTemplate();

    @Bindable
    @Column(nullable = false, length = SmsPatternRestriction.NAME_MAX_LENGTH)
    String getName();

    @Bindable
    @Column(nullable = false, length = SmsPatternRestriction.REGEX_MAX_LENGTH)
    String getRegex();

    @Bindable
    @Column(nullable = false, length = SmsPatternRestriction.SENDER_MAX_LENGTH)
    String getSender();

    @Bindable
    @Nullable
    @Column(name = "date_group")
    Integer getDateGroup();

    @Bindable
    @Nullable
    @Column(name = "value_group")
    Integer getValueGroup();
}

File: app/src/main/java/io/github/zwieback/familyfinance/core/model/ISmsPattern.kt
package io.github.zwieback.familyfinance.core.model

import androidx.databinding.Bindable
import io.github.zwieback.familyfinance.business.sms.handler.SmsHandler
import io.github.zwieback.familyfinance.core.model.restriction.SmsPatternRestriction
import io.requery.*

@Entity(propertyNameStyle = PropertyNameStyle.FLUENT_BEAN, copyable = true)
@Table(name = "t_sms_pattern")
interface ISmsPattern : IBaseEntity {

    @get:Bindable
    @get:ManyToOne(cascade = [CascadeAction.NONE])
    @get:Column(name = "template_id", nullable = false)
    val template: ITemplate

    @get:Bindable
    @get:Column(nullable = false, length = SmsPatternRestriction.NAME_MAX_LENGTH)
    val name: String

    @get:Bindable
    @get:Column(nullable = false, length = SmsPatternRestriction.REGEX_MAX_LENGTH)
    val regex: String

    @get:Bindable
    @get:Column(nullable = false, length = SmsPatternRestriction.SENDER_MAX_LENGTH)
    val sender: String

    @get:Bindable
    @get:Nullable
    @get:Column(name = "date_group")
    val dateGroup: Int?

    @get:Bindable
    @get:Nullable
    @get:Column(name = "value_group")
    val valueGroup: Int?

    /**
     * Required to create a notification about a partially completed operation
     * in the [SmsHandler]
     */
    @get:Bindable
    @get:Column(nullable = false, value = "0")
    val isCommon: Boolean
}
